test/unit/context/test-node-values.js

```diff
@@ -0,0 +1,878 @@
+  });
+
+  function el(id) {
+    if (id == 'T') {
+      return tree;
+    }
+    const found = tree.querySelector(`#${id}`);
+    if (!found) {
+      throw new Error(`element not found ${id}`);
+    }
+    return found;
+  }
+
+  /**
+   * @param {Object} obj
+   * @param {string} name
+   */
+  function wrapper(obj, name) {
+    const original = obj[name];
+    const stub = env.sandbox.stub(ContextNode.prototype, name);
+    const wrapperName = `__wrapper_${name}`;
+    stub.callsFake(function (...args) {
+      const obj = this;
+      const resolvers = obj[wrapperName] ?? (obj[wrapperName] = []);
+      const result = original.apply(this, args);
+      const current = resolvers.slice(0);
+      resolvers.length = 0;
+      current.forEach((resolver) => resolver(result));
+      return result;
+    });
+    return {
+      spy: stub,
+      waitFor: (obj) => {
+        const resolvers = obj[wrapperName] ?? (obj[wrapperName] = []);
+        return new Promise((resolve) => {
+          resolvers.push(resolve);
+        });
+      },
+    };
+  }
+
+  function waitForDiscover(...nodesOrContextNodes) {
+    const contextNodes = nodesOrContextNodes.map((arg) =>
+      arg.nodeType ? ContextNode.get(arg) : arg
+    );
+    const promises = contextNodes.map((contextNode) =>
+      discoverWrapper.waitFor(contextNode)
+    );
+    clock.tick(1);
+    return Promise.all(promises);
+  }
+
+  function rediscover(...nodesOrContextNodes) {
+    const contextNodes = nodesOrContextNodes.map((arg) =>
+      arg.nodeType ? ContextNode.get(arg) : arg
+    );
+    contextNodes.forEach((cn) => cn.discover());
+    return waitForDiscover.apply(null, contextNodes);
+  }
+
+  describe('connected, non-recursive', () => {
+    let sibling1, sibling1Stub;
+    let sibling2, sibling2Stub;
+    let cousin1, cousin1Stub;
+    let parent, parentStub;
+    let grandparent, grandparentStub;
+
+    beforeEach(async () => {
+      doc.body.appendChild(tree);
+      sibling1 = ContextNode.get(el('T-1-1-1'));
+      sibling2 = ContextNode.get(el('T-1-1-2'));
+      cousin1 = ContextNode.get(el('T-1-2-1'));
+      parent = ContextNode.get(el('T-1-1'));
+      grandparent = ContextNode.get(el('T-1'));
+
+      sibling1Stub = sandbox.stub();
+      sibling2Stub = sandbox.stub();
+      cousin1Stub = sandbox.stub();
+      parentStub = sandbox.stub();
+      grandparentStub = sandbox.stub();
+
+      await waitForDiscover(grandparent, parent, sibling1, sibling2, cousin1);
+
+      sibling1.values.subscribe(NonRecursive, sibling1Stub);
+      sibling2.values.subscribe(NonRecursive, sibling2Stub);
+      cousin1.values.subscribe(NonRecursive, cousin1Stub);
+      parent.values.subscribe(NonRecursive, parentStub);
+      grandparent.values.subscribe(NonRecursive, grandparentStub);
+      clock.runAll();
+      calcSpy.resetHistory();
+    });
+
+    it('should resolve with an existing value immediately', async () => {
+      grandparent.values.set(NonRecursive, 'OWNER1', 'A');
```

Comment:
In the design doc I summed the problem up in the "Property setter collisions" section. From the previous work with owners and few other places, this turned out to be the major problem and I wanted to tackle it head on here. To illustrate, something similar happens with CSS as well. Say, you have to reasons why an element could be `display:none`. E.g. it could be hidden because of amp-access rule, or maybe due to a media query, etc. If these rules used `element.style.display = 'none'`, they'd never be able to tell apart the reason. And once, one of these rules (e.g. `amp-access`) allows display, we no longer no if the other rules still apply - hence we get a race condition or a logic problem. Instead, it's safer to separate these rules by using different CSS selectors. E.g.

```
.hide-due-to-amp-access,
.hide-due-to-media,
.hide-due-to-reason1 {
  display: none;
}
```

This way, there's no collision - the element is displayed when there's no longer any rule that prevents it from being displayed. With context properties, it's very similar. From the debugging point of view, this is also very convenient: I can always trace why the property is set the way it is. OTOH, the cost of this is fairly low.